File: java/288. Unique Word Abbreviation.java
public class ValidWordAbbr {
    
    Map< String, Set<String> > mp = new HashMap<>();
    public ValidWordAbbr(String[] dictionary) {
        int len = dictionary.length;
        for(int i = 0; i < len; ++i){
            String s = dictionary[i];
            int sLen = s.length();
            if(s.length() > 2){
                s = s.charAt(0) + Integer.toString(sLen-2) + s.charAt(sLen-1);
            }
            if(mp.containsKey(s)){
                mp.get(s).add(dictionary[i]);
            }else{
                Set<String> set = new HashSet<String>();
                set.add(dictionary[i]);
                mp.put(s, set);
            }
        }
    }

    public boolean isUnique(String word) {
        String s = word;
        int sLen = s.length();
        if(sLen > 2){
           s = s.charAt(0) + Integer.toString(sLen-2) + s.charAt(sLen-1); 
        }
        if(!mp.containsKey(s)) return true;
        else return mp.get(s).contains(word) && mp.get(s).size() <= 1;
    }
}


// Your ValidWordAbbr object will be instantiated and called as such:
// ValidWordAbbr vwa = new ValidWordAbbr(dictionary);
// vwa.isUnique("Word");
// vwa.isUnique("anotherWord");

File: cpp/288. Unique Word Abbreviation.cpp
class ValidWordAbbr {
public:
    unordered_map<string,unordered_set<string>> map; //d2r-> words 
    ValidWordAbbr(vector<string> &dictionary) {
        for(int i = 0; i < dictionary.size(); ++i){
            
            string word = dictionary[i];
            int n = word.length();
            string abbr = word[0] + to_string(n-2)+word[n-1];
            map[abbr].insert(word);
        }
        
    }

    bool isUnique(string word) {
        int n = word.length();
        string abbr = word[0]+to_string(n-2)+word[n-1];
        return map[abbr].count(word) == map[abbr].size();
    }
};



// Your ValidWordAbbr object will be instantiated and called as such:
// ValidWordAbbr vwa(dictionary);
// vwa.isUnique("hello");
// vwa.isUnique("anotherWord");